staging/src/k8s.io/apiserver/pkg/apis/apiserver/types.go

```diff
@@ -175,9 +175,40 @@ type JWTAuthenticator struct {
 	UserValidationRules  []UserValidationRule
 }
 
-// Issuer provides the configuration for a external provider specific settings.
+// Issuer provides the configuration for an external provider's specific settings.
 type Issuer struct {
-	URL                  string
+	// url points to the issuer URL in a format https://url or https://url/path.
+	// This must match the "iss" claim in the presented JWT, and the issuer returned from discovery.
+	// Same value as the --oidc-issuer-url flag.
+	// Used to fetch discovery information unless overridden by discoveryURL.
+	// Required to be unique across all JWT authenticators.
+	// Note that egress selection configuration is not used for this network connection.
+	// +required
+	URL string
+	// discoveryURL if specified, overrides the URL used to fetch discovery information.
+	// /.well-known/openid-configuration will be appended by the authenticator at runtime
+	// to fetch the discovery information.
+	// The "issuer" field in the fetched discovery information must match the "issuer.url" field
+	// in the AuthenticationConfiguration and will be used to validate the "iss" claim in the presented JWT.
+	// This is for scenarios where the well-known and jwks endpoints are hosted at a different
+	// location than the issuer (such as locally in the cluster).
+	// Format must be https://url or https://url/path.
+	//
+	// Example:
+	// A discovery url that is exposed using kubernetes service 'oidc' in namespace 'oidc-namespace'.
+	// certificateAuthority is used to verify the TLS connection and the hostname on the leaf certificate
+	// must be set to 'oidc.oidc-namespace'.
+	//
+	// curl https://oidc.oidc-namespace (.discoveryURL field)
```

Comment:
If discoveryURL is `https://oidc.oidc-namespace`, then with the current documented behavior you would curl `https://oidc.oidc-namespace/.well-known/openid-configuration`, right?

I still think it would be clearer to make discoveryURL the full exact URL we fetch from. If we don't do that, I think we'll be adding a "useDiscoveryURLVerbatim bool" flag in a year to opt out of that behavior. People using discoveryURL are already advanced, giving them full exact control over the url to pull from without assuming the well-known suffix makes sense to me. We can document this very clearly with examples, like this:

```go
	// url points to the issuer URL in a format https://url or https://url/path.
  // ...
	// Discovery information is fetched from "{url}/.well-known/openid-configuration" unless overridden by discoveryURL.
  // ...
	URL string

  // discoveryURL, if specified, overrides the URL used to fetch discovery
  // information instead of using "{url}/.well-known/openid-configuration".
  // The exact value specified is used, so "/.well-known/openid-configuration"
  // must be included in discoveryURL if needed.
  // ...
  DiscoveryURL string
```

Reply:
Made `discoveryURL` the exact URL as per https://github.com/kubernetes/kubernetes/pull/123527#discussion_r1509619257. With this change I've removed the disallow trailing slash in the discovery url check as user has full control over it. LMK what you think!